File: Leetcode/100. Same Tree/Same_Tree.py
# Definition for a binary tree node.

# class TreeNode:

#     def __init__(self, val=0, left=None, right=None):

#         self.val = val

#         self.left = left

#         self.right = right

class Solution:

    def isSameTree(self, p, q):

        """

        :type p: TreeNode

        :type q: TreeNode

        :rtype: bool

        """    

        # p and q are both None

        if not p and not q:

            return True

        # one of p and q is None

        if not q or not p:

            return False

        if p.val != q.val:

            return False

        return self.isSameTree(p.right, q.right) and \

               self.isSameTree(p.left, q.left)

File: Leetcode/100. Same Tree/same_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        
        if( p == NULL && q == NULL){
            return true;
        }
         
         if( p == NULL && q != NULL){
             return false;
         } 

         if(p != NULL && q == NULL){
             return false;
         }  
        
        //recursive call for left part of tree 

        bool left = isSameTree(p->left,q->left);

        //recursive call for right part of tree 

       bool right = isSameTree(p->right,q->right) ; 
       
       // check data;
       bool val = p->val == q->val;

       if(left && right && val)
       {
           return true;
       }
       else{
           return false;
       }

    }
};
